lib/ecto/adapters/sql.ex

```diff
@@ -1351,22 +1398,24 @@ defmodule Ecto.Adapters.SQL do
     end
   end
 
-  defp log_stacktrace(stacktrace, repo) do
-    with [_ | _] <- stacktrace,
-         {module, function, arity, info} <- last_non_ecto(Enum.reverse(stacktrace), repo, nil) do
+  defp log_stacktrace([_ | _] = stacktrace, repo, {module, function, args}) do
+    entries = apply(module, function, args ++ [Enum.reverse(stacktrace), repo])
+
+    for {{module, function, arity, info}, idx} <- Enum.with_index(entries) do
```

Comment:
You could also do `Enum.with_index(entries, fn {module, function, arity, info}, idx ->` so we don't do additional passes on the data.